### ISSUE TITLE
Ettercap filter explicit value

### ISSUE THREAD

**Issue by luismartinsez** [Ettercap/ettercap#942]
I am trying to create an ettercap filter to establish a MITM attack in Modbus TCP traffic.

Could help me to understand what parameters I should be passing in the following "Search" and "Replace"  for changing register 1?

if (ip.proto == TCP && tcp.dst == 502) {
   if (search(DATA.data, "\xE1\x00")) {
    msg("ModbusValue Register 1 Changed from 225 to 3000");
      replace("\xE1\00", "\xBB\x80");

![image](https://user-images.githubusercontent.com/48541300/57187834-a5f1b680-6f27-11e9-987a-1311e68b6c35.png)

When I send the above script, it changes values of all Registers from 1 through to 12, but I need to change the value of only Register 1. Is it possible explicitly define the script to change only Register 1?

**Comment by luismartinsez** [comment]
There's an x missing in previously posted script. Following is the correct one.

if (ip.proto == TCP && tcp.dst == 502) {
if (search(DATA.data, "\xE1\x00")) {
msg("ModbusValue Register 1 Changed from 225 to 3000");
replace("\xE1\x00", "\xBB\x80");

**Comment by koeppea** [comment]
So these two bytes just identify the value, but not the register.

I assume there are a few bytes before that identify the register.

You need to include this in your seach/replace pattern, of course when
replacing, you've to let the bytes to identify register 1 unmodified.

However I'm not aware of the internals of Modbus.

**Comment by koeppea** [comment]
Can you attached (zipped) the packet capture file?

**Comment by luismartinsez** [comment]
Thank you for your response.

I am following a paper published by Gabriel Sanchez of SANS Institute, which is comprehensive. However, the researcher uses a filter to make changes to a single coil register value. In my case, I am trying to change the value of a holding register.

Uploaded pcap & Gabriel’s paper.

ModbusPal can be used to analyse data structure and traffic of Modbus TCP protocol. https://github.com/zeelos/ModbusPal

**Comment by luismartinsez** [comment]
[koeppea.zip](https://github.com/Ettercap/ettercap/files/3146296/koeppea.zip)

**Comment by koeppea** [comment]
So I'm wondering why it even replaced anything.
Because you're only looking at the payload if the TCP destination port is 502.
But actually, the desired content only appears in the response where the TCP port 502 is the source - not the destination.

However, I don't know if this is due to missing curly brackets - but I hardly can't imagine etterfilter would have compiled it.
Try to compile the following filter source file with etterfilter:
```
if (ip.proto == TCP && tcp.src == 502) {
   if (search(DATA.data, "\xE1\x00")) {
      msg("ModbusValue Register 1 Changed from 225 to 3000");
      replace("\xE1\x00", "\xBB\x80");
   }
}
```
Please let me know if this script makes any difference.
If it doesn't help I'll setup a simulation envrionment and replaying the data to see how Ettercap behaves in this testing envrionment.

**Comment by luismartinsez** [comment]
Hi Alexander,

Those two curly brackets did not get copied when I created the ticket, my apologies. They were there in my script. However, as you rightly pointed out, I was pointing to the payload of the destination. Thanks for that - I have corrected it.

I have now used your script that points to the payload of the source. As you can see in the following image, it changed value to 225 for all registers. Something strange!!

![image](https://user-images.githubusercontent.com/48541300/57297821-39331380-7103-11e9-85a3-0e57050f9606.png)

Those holding registers, from 0 to 9, stores incremental values of various processes like Pressure, Temperature, Flow etc. Those values are coming from Programmable Logical Controllers. This means, at some point, each one of the register will have a value of 225, so it is important to instruct the script to only Register 1 and leave other registers unmodified. This is what I am trying to achieve.

Uploaded pcap of today's testing

**Comment by luismartinsez** [comment]
Ooops!! I wrongly pressed Close button

**Comment by luismartinsez** [comment]
[may7.zip](https://github.com/Ettercap/ettercap/files/3152526/may7.zip)

**Comment by koeppea** [comment]
OK I'll have a deeper look.

**Comment by luismartinsez** [comment]
Please let me know if you need any additional information from my side. Thanks Alexander

**Comment by koeppea** [comment]
All set. Just need some free time to setup the test env and reproduce.

**Comment by koeppea** [comment]
Alright,

I've managed to set up  a test environment. Even though I recognized that the byte sequence matches and replaces actually the wrong packets with the wrong replacement values, I made the observation that it's gotten replaced only once.

So effectively I couldn't reproduce the behaviour you've been seeing:

Correct filter been used:
```
if (ip.proto == TCP && tcp.src == 502) {
   if (search(DATA.data, "\x00\xE1")) {
      msg("ModbusValue Register 1 Changed from 225 to 3000");
      replace("\x00\xE1", "\x0B\xB8");
   }
}
```
_Note that 225 == 0x00e1 and 3000 == 0x0bb8._
Find attached a packet capture of a packet containing the payload before and after ettercap filter was applied. The packet capture, where these two sub-captures have been extracted, have been recorded using the `-w` switch while running ettercap.

I've used Ettercap compiled from the latest source from GitHub. So I assume you'd try doing the same.

[ettercap-modbus.zip](https://github.com/Ettercap/ettercap/files/3174861/ettercap-modbus.zip)

The archive also contains the TCP Modbus payload. I used **scapy** to build a packet with the desired payload for reproducing the issue.

**Comment by luismartinsez** [comment]
Hi Alexander,

Appreciate your time and effort in setting up a testbed and providing accurate information. 

I had incorrect hex values in my script and also an older version of ettercap. Used your script and also compiled ettercap latest source from Github. It looks good and replacing the register only once. Thanks

Can you think of an idea to identify Register 1 and change value of this register only? This is because the script can modify the value of other Registers from 2 through to 9, when it finds a value of 255, and not necessarily Register 1. So, it is important to instruct the script to only change payload of Register 1 and leave other registers unmodified. Is this possible?

**Comment by koeppea** [comment]
I think the Modbus header is of fixed size.
I assume:
2 bytes transaction ID
2 bytes protocol ID
2 bytes length
1 byte unit identifier
That's for the Modbus/TCP header. Then you are at the Modbus Part:
1 byte flags
1 byte Byte Count
2 bytes 1st Register (0-register)
2 bytes 2nd Register (1-register)
.... until the 10th (9-register) since the byte count says when it's end.

So described as a filter I'd say (reading the _etterfilter_ man-page):
```
if (tcp.src == 502 && DATA.data + 9 == "\x00\xe1") {
   DATA.data + 9 = "\x0b\xb8";
}
```

**Comment by luismartinsez** [comment]
Thanks Alexander,
This script AS IS not doing anything. I cannot find the value anywhere.

You mean to say to include this logic into the earlier search & find script?

**Comment by luismartinsez** [comment]
Are you able to successfully change this on your testbed, using this script?

**Comment by luismartinsez** [comment]
Hi Alexander,

I spent a few hours without much luck.

Is there a way to print the value of DATA.data + 9 to ensure we're looking at Register 1? I tried the following, but it doesn't want to compile.

if (tcp.src == 502 && DATA.data + 9 == "\x00\xe1") 
msg(eval(DATA.data + 9));
{ DATA.data + 9 = "\x0b\xb8"; }

**Comment by koeppea** [comment]
Didn’t had the time yet to look at it

**Comment by luismartinsez** [comment]
Thought so, no problem :-)

**Comment by koeppea** [comment]
OK I investigated the issue and I understand now why it's not working as it's supposed to work.
The following filter works though:
```
if (tcp.src == 502 && DATA.data+12 == "\xe1") {
   DATA.data + 11 = "\x0b\xb8";
   msg("Data replaced");
}
```
This looks kinda weird but if you understand how string operations work it becomes clear.
For reference: **DATA.data** looks like the following:
```
00000000  38 33 00 00 00 17 00 03  14 00 00 00 e1 00 00 08  |83..............|
00000010  ca 3d 86 fa 7d ff ff ff  ff ff ff ff ff           |.=..}........|
```
I'm matching at offset 12 even though the correct value for register index 1 begins at offset 11.

Apparently _etterfilter_ translates the above condition **DATA.data+12** into: `strdup(DATA.data+12)` before the actual comparsion is being performed. 

If we would move the pointer to the correct position where register index 1 really starts (**DATA.data + 11**: `0x00e1`) it becomes `strdup("\x00\xe1")`. The leading zero byte terminates the string to be copied too early, so that the condition doesn't match as it would come to the comparsion `"\x00" ==  "\x00\xe1"` which is not equal. So the condition of the **if** clause isn't satisfied and the actual replacement isn't performed at all.

I wanna try some things as I think the key is to use `memcpy()` instead. But _etterfilter_ is tricky, so I've to investigate a bit further.

P.S.
The offset of 9 in my last suggested filter was not totally wrong but wrong for your case. It points at register index 0 which is actually the first register. However from the example payload provided you intend to replace register index 1 which is the second register actually.

P.P.S
`eval()` is not yet  a valid function supported by _etterfilter_. However I think it's worth realizing such a function as I've also missed such kind of function.... putting it onto the wishlist.

**Comment by luismartinsez** [comment]
Fantastic Alexander - brilliant observation !!

You admitted initially that you're unfamiliar with this protocol, but you jumped in, confident in your own determination and succeeded. Great!!

**Comment by koeppea** [comment]
Just for caution: the assign operator **=** has the same weakness or issue. So if you want to assign a value of `0x00ab` on offset 11, you have to assign `"\x00"` at offset 11 and in a separate assignment `"\xab"` at offset 12. Otherwise everything that follows the zero byte is gibberish which is typical for uninitialized memory.
When a zero-byte is involved you've to take care until this issue is fixed, which will take me some time.

**Comment by koeppea** [comment]
Hi @luismartinsez 

I've pushed a pull request to fix the issue of **etterfilter**.
I've tested the filter script which was not working (or only with the offset tricks) before.
With the changes in #945 applied, the condition matches regardless if binary coded zero-bytes are contained or not.
The assignment also works also for values with leading or intermediate zero-bytes.

The following filter works with the code of #945:
```
if (tcp.src == 502 && DATA.data+11 == "\x00\xe1") {
   DATA.data + 11 = "\x08\xb0";
   msg("Data replaced");
}
```

Note that you've to recompile **ettercap** with this pull request and also recompile the filter after **ettercap** has been built.

Please check if the pull request also fixes the issue for you.
Let me know if you need help to do that test.

**Comment by luismartinsez** [comment]
Glad to know you have fixed it!! Thanks Alexander

Could you help me to "recompile ettercap with this pull request" please?. Do I have to first remove ettercap completely from Ubuntu 18.04.2 LTS? I am not too familiar with pull request in github and compile. Could you send me instructions please?

**Comment by koeppea** [comment]
Yes you need to purge ettercap first: `apt-get purge ettercap`.
Then you need to install the following dependencies to build ettercap from source (extracted from **README**):
```
apt-get install build-essential git
apt-get install debhelper bison check cmake flex ghostscript libbsd-dev \
      libcurl4-openssl-dev libgeoip-dev libltdl-dev libluajit-5.1-dev \
      libncurses5-dev libnet1-dev libpcap-dev libpcre3-dev libssl-dev \
      libgtk-3-dev libgtk2.0-dev 
```

Then you need to clone the my repository, and checkout to the branch for fixing the issue:
```
git clone https://github.com/koeppea/ettercap
cd ettercap
git checkout -b koeppea-fix-942
git merge origin/fix-942
```

Then create  a build directory and build and install ettercap:
```
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Debug ..
make
sudo make install
```

There may be something going wrong during the **cmake** which is normally when a dependency is still missing. In this case you can install the missing dependency and  continue from the **cmake** step.

HTH

**Comment by luismartinsez** [comment]
Thanks Alexander, it worked :-) and successfully replaced the value in position 11. Great job!!!

A couple of questions:

a. Now we know that it successfully changes the value of Register 1, which is in position 11. What would be the position of Register 0, Register 2, Register 3... through to Register 9? How do I find these positions?

I tried running the following, but it's not going to the second if block.

if (tcp.src == 502 && DATA.data+11 == "\x00\xe1") {
  DATA.data + 11 = "\x0b\xb8";
   msg("Holding Register 1 data replaced");
}

if (tcp.src == 502 && DATA.data+10 == "\x00\x00") {
   DATA.data + 10 = "\x00\x7d";
   msg("Holding Register 0 replaced");
}
if (tcp.src == 502 && DATA.data+12 == "\x00\x00") {
   DATA.data + 12 = "\x00\x7d";
   msg("Holding Register 2 data replaced");
}

b. I observed that the attached script ( that has multiple if conditions) doesn't want to work. It used to work in the older version.

c. Also, script doesn't work if there are more than 1 target. I mean it's not looking for data in target 2, target 3 etc

Can you please have a look? 

[scen.zip](https://github.com/Ettercap/ettercap/files/3227496/scen.zip)

**Comment by koeppea** [comment]
You can analyse the packets of interesst with Wireshark before. Then you need to figure out the offset from the TCP payload to where the 1st register (0) begins. Luckily Wireshark is already capable to dissect the Modbus protocol.

So remember how did the example payload looks like which you've provided:
```
00000000  38 33 00 00 00 17 00 03  14 00 00 00 e1 00 00 08  |83..............|
00000010  ca 3d 86 fa 7d ff ff ff  ff ff ff ff ff           |.=..}........|
```

So the first register (0) begins at offset 9. Each Register is 2 bytes long.
How many registers are available can be found in byte 8 just before the registers begin: (0x14 ==  20 bytes). This makes 10 registers á 2 bytes.

b)
The main issue is, that you didn't honor the length of a register. Each is 2 bytes. So you have to increase the offset by 2 starting from the first register.

However your conditions didn't match because you've modified the content you tried to match afterwards:
Have a look at your filter script:
> ```
> if (tcp.src == 502 && DATA.data+11 == "\x00\xe1") {
> DATA.data + 11 = "\x0b\xb8";
> msg("Holding Register 1 data replaced");
> }
> ```

If the payload + 11 bytes matches 0x00e1, you replace the payload at offset 11 with 0x0bb8.
So offset 11 has 0x0b and offset 12 has 0xb8 as you replaced with 2 bytes.
Next condition....
> ```
> if (tcp.src == 502 && DATA.data+10 == "\x00\x00") {
> DATA.data + 10 = "\x00\x7d";
> msg("Holding Register 0 replaced");
> }
> ```

This condition cannot match 0x0000 at offset 10, because you've already modified offset 11 to 0x0b. So you find now at offset 10 0x000b.
Next condition...
> ```
> if (tcp.src == 502 && DATA.data+12 == "\x00\x00") {
> DATA.data + 12 = "\x00\x7d";
> msg("Holding Register 2 data replaced");
> }
> ```

This condition doesn't match, because at offset 12 there is no 0x0000.

c)
What do you mean with different targets? I've resent the payload to different IPs and the filter matched, disregarding which IP.
The followng worked for me:

```
#
# Replace Register 0 only if it contains 0x0000
#
if (tcp.src == 502 && DATA.data+9 == "\x00\x00") {
   DATA.data + 9 = "\xde\xad";
   msg("Data in register 0 replaced");
}


#
# Replace Register 1 only if it contains 0x00e1
#
if (tcp.src == 502 && DATA.data+11 == "\x00\xe1") {
   DATA.data + 11 = "\xbe\xef";
   msg("Data in register 1 replaced");
}


#
# Replace Register 2 regardless of it's current value
#
if (tcp.src == 502) {
   DATA.data + 13 = "\xc0\xfe";
   msg("Data in register 2 replaced");
}


#
# Replace Register 3 regardless of it's current value
#
if (tcp.src == 502) {
   DATA.data + 15 = "\xca\xfe";
   msg("Data in register 3 replaced");
}
```

```
Text only Interface activated...
Hit 'h' for inline help

Data in register 0 replaced
Data in register 1 replaced
Data in register 2 replaced
Data in register 3 replaced
Closing text interface...
```
Original Payload:
![image](https://user-images.githubusercontent.com/4201853/58512131-9e2ad800-819c-11e9-840b-d0233767c64f.png)

Replaced Payload:
![image](https://user-images.githubusercontent.com/4201853/58512345-2e691d00-819d-11e9-97e4-5c24c605441e.png)


BTW: the attached ZIP contains a different filter than you mentioned in your last  post.

**Comment by luismartinsez** [comment]
Thanks Alexander for your detailed explanation, as usual.
I have tested with various scenarios and scripts. Everything is working as expected. 

Thanks for getting this done. Keep your positive attitude alive, and never let the enthusiasm in you fade, this is your strength. I will create a ticket if I need anymore help in future :-)

**Comment by koeppea** [comment]
#945 merged. Hence closing this issue as resolved.

**Comment by salwaalem** [comment]
Hello Alexander,

Please I need your help.

When I compile my Modbus.filter which contains that:

if (tcp.src == 502 && DATA.data + 97 == "\x00\x07") {

DATA.data + 97 = "\x00\x00";
msg("Data replaced");
}

With the command line :

/usr/share/ettercap# ettercap Modbus.filter

I have the below message:

ettercap 0.8.2 copyright 2001-2015 Ettercap Development Team

Any idea please ?

**Comment by koeppea** [comment]
@salwaalem this way the filter is not compiled.
Read the man-page for **etterfilter**.
Even though the file extension doesn't matter at all, it has become common sense that the filter source files have the extension _ecf_ and the compiled filter _ef_.
So a common expression to compile a filter would be:
```
etterfilter -o /tmp/filter.ef filter.ecf
```

**Comment by salwaalem** [comment]
Thank you very much it's compiled without problem.
I'll see if it will change the right register or not....

**Comment by salwaalem** [comment]
Hi,

I have a problem with ettercap filter. When I put my filter It makes too much time to replace data 
And I have a warning my console. 
My filter is this one: 
if (tcp.src == 502 && DATA.data + 11 == "\x00\x00") {

DATA.data + 11 = "\x00\x01";
msg("Data replaced");
}

![ettercap2](https://user-images.githubusercontent.com/40567501/73563795-7d6afe80-445e-11ea-80d9-b8dcf77516c5.PNG)

![ettercap](https://user-images.githubusercontent.com/40567501/73563724-58768b80-445e-11ea-89e1-18899fec6641.PNG)

Any help please ?

**Comment by koeppea** [comment]
@salwaalem: the fix that fixed this issue, isn't included in the version you're running.
So please first update to the latest source code version.
See [here](https://github.com/Ettercap/ettercap/wiki/Providing-debug-information) how to do that.
If it's still not working, follow the above mentioned instructions in detail and open a new issue for your problem.

**Comment by salwaalem** [comment]
Thanks Alexander. 
I followed the mentionned instruction but it's still not working with a new critical error. I will open a new issue
